### DISCUSSION TITLE
Extracting Error codes

### DISCUSSION THREAD

**Discussion by anwang2009** [rust-db/refinery#142] [Q&A]
Hi, I noticed the possible error kinds are

```
pub enum Kind {
    InvalidName,
    InvalidVersion,
    DivergentVersion(Migration, Migration),
    MissingVersion(Migration),
    InvalidMigrationPath(PathBuf, std::error::Error),
    ConfigError(String),
    Connection(String, Box<dyn std::error::Error + Sync + Send>),
}
```

From reading the codebase, afaict more specific errors such as `tokio_postgres::error::Error` [(docs)](https://docs.rs/tokio-postgres/0.6.0/tokio_postgres/error/struct.Error.html) are swallowed up as std::error::Error in Kind::Connection, so there is no way of extracting error codes. Is that correct?

**Comment by jxs** [comment]
Hi there, you can access the underlying error by calling [source](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) on `refinery::Error`

**Reply by anwang2009** [reply]
I see, I am trying to avoid doing a string parse for an error code. I don't think I can get around the parse if `source` is all I've got for now. thanks!

**Reply by jxs** [reply]
you don't need to string parse, you can do something like:
```Rust
use rusqlite::{Connection, Error as SqliteError};
use std::error::Error;

mod embedded {
    use refinery::embed_migrations;
    embed_migrations!("../refinery/refinery/tests/sql_migrations");
}

#[tokio::main]
async fn main() {
    let mut conn = Connection::open_in_memory().unwrap();
    match embedded::migrations::runner().run(&mut conn) {
        Ok(_) => {}
        Err(err) => {
            if let Some(err) = err
                .source()
                .map(|source| source.downcast_ref::<SqliteError>())
            {
            // handle err
            }
        }
    }
}

```
